packages/astro/src/core/ssr/index.ts

```diff
@@ -160,6 +160,12 @@ export async function ssr({ astroConfig, filePath, logging, mode, origin, pathna
 
     let html = await renderPage(result, Component, pageProps, null);
 
+    // inject <!doctype html> if missing (TODO: is a more robust check needed for comments, etc.?)
+
+    if (!/<!DOCTYPE html/i.test(html)) {
+      html = '<!DOCTYPE html>\n' + html;
```

Comment:
From our meeting this morning, this can be moved to the compiler eventually. But a RegEx + insertion here takes ~`0.05ms` so it’ll hold us over for now.

(for comparison, parsing the full doc HTML with htmlparser2 takes ~`0.15ms` which is more correct but I didn’t know if it was necessary for now)